A4A: Format currency

yashwin opened the pull request:
Resolves https://github.com/Automattic/i18n-issues/issues/921
Resolves https://github.com/Automattic/automattic-for-agencies-dev/issues/1833

## Proposed Changes

* This PR fixes the currency formatting issues across A4A.

## Testing Instructions

* Open the A4A live link
* Verify the current formatting works as expected at the mentioned places:

You can see the currency will have US$ instead of $ if you are based out of the US.

| Before | After |
|--------|--------|
| <img width="897" alt="Screenshot 2025-02-24 at 10 19 09 AM" src="https://github.com/user-attachments/assets/f59063c0-09b6-4280-b841-6cbbab6a0861" /> |  <img width="897" alt="Screenshot 2025-02-24 at 10 19 01 AM" src="https://github.com/user-attachments/assets/f3316911-8afe-46eb-8f38-652b0b13b179" />| 

| Before | After |
|--------|--------|
| <img width="897" alt="Screenshot 2025-02-24 at 10 19 38 AM" src="https://github.com/user-attachments/assets/a534743f-640f-4792-82e9-f85412344ca1" /> | <img width="897" alt="Screenshot 2025-02-24 at 10 19 35 AM" src="https://github.com/user-attachments/assets/6be7a6ab-d09d-46d5-b6e8-887484652459" /> | 

| Before | After |
|--------|--------|
| <img width="897" alt="Screenshot 2025-02-24 at 10 19 59 AM" src="https://github.com/user-attachments/assets/acf82cdc-c8ae-4a34-9f6f-376955d70110" /> | <img width="904" alt="Screenshot 2025-02-24 at 10 54 33 AM" src="https://github.com/user-attachments/assets/7a221022-0369-481d-87bb-cd424bc3454b" /> | 

| Before | After |
|--------|--------|
| <img width="782" alt="Screenshot 2025-02-24 at 10 23 07 AM" src="https://github.com/user-attachments/assets/3e2be50c-8321-4065-ad3d-d7415acb9ef7" /> | <img width="782" alt="Screenshot 2025-02-24 at 10 23 10 AM" src="https://github.com/user-attachments/assets/577d3d82-a776-4871-bf33-9147d189e35d" /> | 

| Before | After |
|--------|--------|
| <img width="925" alt="Screenshot 2025-02-24 at 10 24 02 AM" src="https://github.com/user-attachments/assets/3dd9ba01-6ad2-4e0b-93eb-0f7d3a6da73b" /> |  <img width="925" alt="Screenshot 2025-02-24 at 10 23 42 AM" src="https://github.com/user-attachments/assets/140e0694-3334-4c09-88bf-8dc181633142" /> | 

| Before | After |
|--------|--------|
| <img width="1136" alt="Screenshot 2025-02-24 at 10 26 17 AM" src="https://github.com/user-attachments/assets/448fd639-fe04-4ff3-8e0e-951dd3be7824" /> | <img width="1136" alt="Screenshot 2025-02-24 at 10 26 11 AM" src="https://github.com/user-attachments/assets/d519f95e-46ba-4bd8-b6a6-6ffb85942a0b" /> | 

| Before | After |
|--------|--------|
| <img width="1339" alt="Screenshot 2025-02-24 at 10 29 02 AM" src="https://github.com/user-attachments/assets/2b95494e-37fd-4f70-b538-64835bdfe715" />  |  <img width="1339" alt="Screenshot 2025-02-24 at 10 28 54 AM" src="https://github.com/user-attachments/assets/265ee36c-d1e0-4444-a088-eb342da35997" /> | 

| Before | After |
|--------|--------|
| <img width="734" alt="Screenshot 2025-02-24 at 10 32 30 AM" src="https://github.com/user-attachments/assets/129faca9-9fb8-4fdc-9766-df77e4e27e38" /> | <img width="734" alt="Screenshot 2025-02-24 at 10 32 37 AM" src="https://github.com/user-attachments/assets/4be118fa-1475-4311-9c5c-cbeb8b595b45" /> | 

| Before | After |
|--------|--------|
| <img width="762" alt="Screenshot 2025-02-24 at 10 33 56 AM" src="https://github.com/user-attachments/assets/9c453282-5973-48c0-b331-72798b350694" /> | <img width="762" alt="Screenshot 2025-02-24 at 10 33 52 AM" src="https://github.com/user-attachments/assets/9ff65b84-da88-478f-b9e3-e7c1fa459a72" /> | 

| Before | After |
|--------|--------|
|  <img width="762" alt="Screenshot 2025-02-24 at 10 51 04 AM" src="https://github.com/user-attachments/assets/7cced9be-43b7-40ec-9015-1c53cbeb048e" /> | <img width="762" alt="Screenshot 2025-02-24 at 10 50 30 AM" src="https://github.com/user-attachments/assets/04ada963-52cd-4324-895b-846b67dcc1cf" /> | 

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

yashwin committed:
A4A: Format currency

yashwin commented on client/a8c-for-agencies/sections/referrals/primary/commission-overview/index.tsx:
```diff
@@ -99,7 +100,13 @@ export default function CommissionOverview( {
+									" For example, if your client's store generates %(maxAmount)s in TPV per year, your revenue share for that year would be %(amount)s.",
+								{
+									args: {
+										maxAmount: formatCurrency( 1000000, 'USD' ),
```
Probably we can stick to $1M here?

<img width="725" alt="Screenshot 2025-02-21 at 3 28 04 PM" src="https://github.com/user-attachments/assets/61ab3b5a-ecb7-4ea6-85e4-5b169a43d505" />

chriskmnds replied:
Interesting! We should probably add a "compact" notation to `formatCurrency`. Let's stick with `$1M` and I can follow up on this separately. 

cc @sirbrillig / @Automattic/i18n  We can just allow `formatCurrency` to accept `numberFormatOptions` in the same way `numberFormat` does, so someone can just tailor the rendering a little better. Or otherwise we can create a `numberFormatCompactCurrency` abstraction. I'll create an issue.

yashwin replied:
Thanks!

I think it's better to format this the same way. I did a workaround: https://github.com/Automattic/wp-calypso/pull/100135/commits/bd67c53b68c08057d034647a154f22f09957a2ac, a quick hack to get this right. Let me know if this works until we have a function for it. Or we could skip the formatting for this particular section for both $1M and $500.

<img width="775" alt="Screenshot 2025-02-21 at 5 14 06 PM" src="https://github.com/user-attachments/assets/442820ce-7158-4ef2-9ff1-eebefe415e51" />
<img width="775" alt="Screenshot 2025-02-21 at 5 18 51 PM" src="https://github.com/user-attachments/assets/11031fe0-4752-4217-a5b3-7c417d176e60" />

chriskmnds replied:
This is perfect. Thanks! Good use of the currency object parts. Glad to see there is a workaround with the existing APIs :-)

yashwin replied:
Awesome. I created a function so that we can easily replace all later: https://github.com/Automattic/wp-calypso/pull/100135/commits/fb2629d9de1e356ca02e886206c948f83e0e062b

yashwin commented on client/a8c-for-agencies/sections/partner-directory/components/budget-selector.tsx:
```diff
@@ -14,27 +15,39 @@ const BudgetSelector = ( { setBudget, budgetLowerRange }: Props ) => {
 		},
 		{
-			label: '$500',
+			label: formatCurrency( 500, 'USD', {
```
EL & EN:

<img width="795" alt="Screenshot 2025-02-21 at 3 20 33 PM" src="https://github.com/user-attachments/assets/015e2c14-0aa2-4f18-a696-2c081a8c9519" />

<img width="804" alt="Screenshot 2025-02-21 at 3 21 06 PM" src="https://github.com/user-attachments/assets/c240b089-f90a-437f-9dae-5d495c40cf45" />

yashwin commented on client/a8c-for-agencies/components/a4a-migration-offer-v3/index.tsx:
```diff
@@ -47,11 +48,16 @@ const MigrationOfferV3 = ( { isExpanded, onToggleView }: Props ) => {
 					<span>
 						{ translate(
-							'{{b}}Limited time offer:{{/b}} Migrate your sites to Pressable or WordPress.com and earn up to $10,000!*',
+							'{{b}}Limited time offer:{{/b}} Migrate your sites to Pressable or WordPress.com and earn up to %(maxCommission)s*',
```
EL & EN:

<img width="1428" alt="Screenshot 2025-02-21 at 3 30 14 PM" src="https://github.com/user-attachments/assets/ed560f67-f51e-4f9a-80e0-15c4f598e349" />


<img width="1428" alt="Screenshot 2025-02-21 at 3 29 53 PM" src="https://github.com/user-attachments/assets/204db064-892a-4c64-8d0d-c5910db8e6ba" />

chriskmnds commented on client/a8c-for-agencies/sections/marketplace/hosting-overview-v3/hosting-content/premier-agency-hosting/pressable-plan-section/index.tsx:
```diff
@@ -245,7 +246,13 @@ export default function PressablePlanSection( {
 				<span className="pressable-plan-section__details-footnote">
 					{ translate(
-						`*If you exceed your plan's storage or traffic limits, you will be charged $0.50 per GB and $8 per 10K visits per month.`
+						`*If you exceed your plan's storage or traffic limits, you will be charged %(storageCharge)s per GB and %(trafficCharge)s per 10K visits per month.`,
```
> per 10K visits

I think let's also pass `10K` through `numberFormatCompact` - it should render as `10 χιλ.` in EL/Greek.

chriskmnds reviewed (commented):
This is awesome. There were quite a few after all 👍 

Thanks for spotting all of these cases. I left some comments, and I'll follow up myself on one of them. Thanks for bringing it up.

yashwin committed:
more fixes

yashwin committed:
more formatting

yashwin committed:
create formatCurrencyCompact

yashwin committed:
prettify text

yashwin commented:
Thanks for the initial review, @chriskmnds! 🙇 

I have updated the PR description with the testing instructions, and it is now ready for a final review. Could you please take a final look? I have also added the `String Freeze` label

chriskmnds commented on client/a8c-for-agencies/lib/currency.ts:
It's awesome that we are thinking about abstracting this way. My only concern is where this lives and if it could lead to other devs importing and using from here. I'd rather if we can avoid having disparate abstractions in the code, rather centralize everything so we don't run into the all sorts of issues that we faced recently. Ideally, this should live under `i18n-calypso` package (until we create a separate number-formatters package). I am currently working on bringing everything into `i18n-calypso` (this PR brings `formatCurrency`: https://github.com/Automattic/wp-calypso/pull/100097 and immediately after we will bring `getCurrencyObject`). So the ideal scenario is for this abstraction to be built into `i18n-calypso` as soon as these two are ready.

So we'll need a follow-up here. 🤔 

Can we maybe rename the function as `__DO_NOT_IMPORT__formatCurrencyCompact`? as done in https://github.com/Automattic/wp-calypso/pull/100097. Also let's add a comment that we are in the process of migrating everything to `i18n-calypso` and reference the above PR (https://github.com/Automattic/wp-calypso/pull/100097). I'm not sure what else will work that's also not adding more to your plate. WDYT? :-)

yashwin replied:
Makes sense. For now, I moved it back to the component: https://github.com/Automattic/wp-calypso/pull/100135/commits/b4f1b337123a076804032e09099c85dc26091b24

chriskmnds commented on client/a8c-for-agencies/sections/referrals/primary/commission-overview/index.tsx:
```diff
@@ -35,6 +37,11 @@ export default function CommissionOverview( {
 		: translate( 'Referrals - Commission details and terms' );
 
+	// TODO: This is a workaround to keep the formatting of the max amount consistent until
+	// we can use the new formatCurrency function that gives the compact number in the correct format.
```
Nice! Yes, this is the sort of comment I was referring to above for the function doc :-)

chriskmnds commented:
Really cool. Thanks again for the incredible work and for being so thorough 😎

chriskmnds reviewed (approved):
Really cool. Thanks again for the incredible work and for being so thorough 😎 

Left a NIT comment, just to help us come back to easier for the follow-up.

yashwin committed:
move formatCurrencyCompact inside the component

yashwin commented:
> Really cool. Thanks again for the incredible work and for being so thorough 😎
> 
> Left a NIT comment, just to help us come back to easier for the follow-up.

Thanks for the review, @chriskmnds! 🙇 

I have updated the PR

jkguidaven reviewed (approved):
Tested and it works well. Changes LGTM! 👍

yashwin committed:
use formatCurrency from i18n-calypso

yashwin committed:
strip zeroes

yashwin committed:
use formatCurrency from i18n-calypso

yashwin committed:
minor import fix
